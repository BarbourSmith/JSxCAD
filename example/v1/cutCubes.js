import '@jsxcad/api-v1-pdf';

Box(30)
  .add(Box(30).move(5, 5), Box(30).move(-5, -5))
  .section()
  .item()
  .Page()
  .view()
  .writePdf('cutCubes');
