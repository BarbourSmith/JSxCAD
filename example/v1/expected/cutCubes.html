<html>
  <head>
    <title>JSxCAD Viewer</title>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0"
    />
  </head>
  <body>
    <script type="module">
      import { toDomElement } from 'https://gitcdn.link/cdn/jsxcad/JSxCAD/master/es6/jsxcad-ui-notebook.js';

      const notebook = [
        {
          view: {
            width: 512,
            height: 256,
            position: [100, -100, 100],
            geometry: {
              type: 'layout',
              layout: {
                size: [42.00000000000001, 42],
                margin: 1,
                title: '42 x 42 : undefined',
              },
              marks: [
                [0, 0, 0],
                [42.00000000000001, 42, 0],
              ],
              content: [
                {
                  type: 'item',
                  content: [
                    {
                      type: 'disjointAssembly',
                      content: [
                        {
                          type: 'disjointAssembly',
                          content: [
                            {
                              type: 'item',
                              tags: ['item/undefined'],
                              content: [
                                {
                                  type: 'layers',
                                  content: [
                                    {
                                      type: 'z0Surface',
                                      tags: ['item/undefined'],
                                      z0Surface: [
                                        [
                                          [-15.000000000000004, 10, 0],
                                          [-20, 10, 0],
                                          [
                                            -20.000000000000004,
                                            -19.999999999999993, 0,
                                          ],
                                        ],
                                        [
                                          [
                                            -20.000000000000004,
                                            -19.999999999999993, 0,
                                          ],
                                          [9.999999999999996, -20, 0],
                                          [
                                            9.999999999999996,
                                            -14.999999999999998, 0,
                                          ],
                                        ],
                                        [
                                          [
                                            9.999999999999996,
                                            -14.999999999999998, 0,
                                          ],
                                          [14.999999999999996, -15, 0],
                                          [
                                            14.999999999999996,
                                            -10.000000000000004, 0,
                                          ],
                                        ],
                                        [
                                          [
                                            14.999999999999996,
                                            -10.000000000000004, 0,
                                          ],
                                          [
                                            20.000000000000004,
                                            -10.000000000000005, 0,
                                          ],
                                          [19.999999999999996, 20, 0],
                                        ],
                                        [
                                          [19.999999999999996, 20, 0],
                                          [-10.000000000000004, 20, 0],
                                          [
                                            -10.000000000000004,
                                            15.000000000000007, 0,
                                          ],
                                        ],
                                        [
                                          [
                                            -10.000000000000004,
                                            15.000000000000007, 0,
                                          ],
                                          [-15.000000000000004, 15, 0],
                                          [-15.000000000000004, 10, 0],
                                        ],
                                        [
                                          [-15.000000000000004, 10, 0],
                                          [
                                            -20.000000000000004,
                                            -19.999999999999993, 0,
                                          ],
                                          [
                                            9.999999999999996,
                                            -14.999999999999998, 0,
                                          ],
                                        ],
                                        [
                                          [
                                            9.999999999999996,
                                            -14.999999999999998, 0,
                                          ],
                                          [
                                            14.999999999999996,
                                            -10.000000000000004, 0,
                                          ],
                                          [19.999999999999996, 20, 0],
                                        ],
                                        [
                                          [19.999999999999996, 20, 0],
                                          [
                                            -10.000000000000004,
                                            15.000000000000007, 0,
                                          ],
                                          [-15.000000000000004, 10, 0],
                                        ],
                                        [
                                          [-15.000000000000004, 10, 0],
                                          [
                                            9.999999999999996,
                                            -14.999999999999998, 0,
                                          ],
                                          [19.999999999999996, 20, 0],
                                        ],
                                      ],
                                    },
                                  ],
                                },
                              ],
                            },
                          ],
                        },
                      ],
                    },
                  ],
                },
                {
                  type: 'disjointAssembly',
                  content: [
                    {
                      type: 'disjointAssembly',
                      content: [
                        {
                          type: 'disjointAssembly',
                          content: [
                            {
                              type: 'paths',
                              tags: ['compose/non-positive', 'color/red'],
                              paths: [
                                [
                                  [21.000000000000004, 21, 0],
                                  [-21.000000000000004, 21.000000000000004, 0],
                                  [-21.000000000000007, -20.999999999999996, 0],
                                  [21, -21.000000000000007, 0],
                                ],
                              ],
                            },
                          ],
                        },
                        {
                          type: 'paths',
                          tags: ['compose/non-positive', 'color/red'],
                          paths: [
                            [
                              null,
                              [-19.781250000000004, 24.28125, 0],
                              [-20.718750000000004, 22.96875, 0],
                              [-19.312500000000004, 22.96875, 0],
                            ],
                            [
                              null,
                              [-19.781250000000004, 24.28125, 0],
                              [-19.781250000000004, 22.3125, 0],
                            ],
                            [null],
                            [
                              null,
                              [-18.750000000000004, 23.8125, 0],
                              [-18.750000000000004, 23.90625, 0],
                              [-18.656250000000004, 24.09375, 0],
                              [-18.562500000000004, 24.1875, 0],
                              [-18.375000000000004, 24.28125, 0],
                              [-18.000000000000004, 24.28125, 0],
                              [-17.812500000000004, 24.1875, 0],
                              [-17.718750000000004, 24.09375, 0],
                              [-17.625000000000004, 23.90625, 0],
                              [-17.625000000000004, 23.71875, 0],
                              [-17.718750000000004, 23.53125, 0],
                              [-17.906250000000004, 23.25, 0],
                              [-18.843750000000004, 22.3125, 0],
                              [-17.531250000000004, 22.3125, 0],
                            ],
                            [null],
                            [null],
                            [
                              null,
                              [-15.468750000000004, 23.625, 0],
                              [-14.437500000000004, 22.3125, 0],
                            ],
                            [
                              null,
                              [-14.437500000000004, 23.625, 0],
                              [-15.468750000000004, 22.3125, 0],
                            ],
                            [null],
                            [null],
                            [
                              null,
                              [-11.437500000000002, 24.28125, 0],
                              [-12.375000000000002, 22.96875, 0],
                              [-10.968750000000002, 22.96875, 0],
                            ],
                            [
                              null,
                              [-11.437500000000002, 24.28125, 0],
                              [-11.437500000000002, 22.3125, 0],
                            ],
                            [null],
                            [
                              null,
                              [-10.406250000000002, 23.8125, 0],
                              [-10.406250000000002, 23.90625, 0],
                              [-10.312500000000002, 24.09375, 0],
                              [-10.218750000000002, 24.1875, 0],
                              [-10.031250000000002, 24.28125, 0],
                              [-9.656250000000002, 24.28125, 0],
                              [-9.468750000000002, 24.1875, 0],
                              [-9.375000000000002, 24.09375, 0],
                              [-9.281250000000002, 23.90625, 0],
                              [-9.281250000000002, 23.71875, 0],
                              [-9.375000000000002, 23.53125, 0],
                              [-9.562500000000002, 23.25, 0],
                              [-10.500000000000002, 22.3125, 0],
                              [-9.187500000000002, 22.3125, 0],
                            ],
                            [null],
                            [null],
                            [
                              null,
                              [-6.937500000000002, 23.625, 0],
                              [-7.031250000000002, 23.53125, 0],
                              [-6.937500000000002, 23.4375, 0],
                              [-6.843750000000002, 23.53125, 0],
                              [-6.937500000000002, 23.625, 0],
                            ],
                            [
                              null,
                              [-6.937500000000002, 22.5, 0],
                              [-7.031250000000002, 22.40625, 0],
                              [-6.937500000000002, 22.3125, 0],
                              [-6.843750000000002, 22.40625, 0],
                              [-6.937500000000002, 22.5, 0],
                            ],
                            [null],
                            [null],
                            [
                              null,
                              [-4.59375, 23.625, 0],
                              [-4.59375, 22.6875, 0],
                              [-4.5, 22.40625, 0],
                              [-4.3125, 22.3125, 0],
                              [-4.03125, 22.3125, 0],
                              [-3.84375, 22.40625, 0],
                              [-3.5625, 22.6875, 0],
                            ],
                            [null, [-3.5625, 23.625, 0], [-3.5625, 22.3125, 0]],
                            [null],
                            [null, [-2.8125, 23.625, 0], [-2.8125, 22.3125, 0]],
                            [
                              null,
                              [-2.8125, 23.25, 0],
                              [-2.53125, 23.53125, 0],
                              [-2.34375, 23.625, 0],
                              [-2.0625, 23.625, 0],
                              [-1.875, 23.53125, 0],
                              [-1.78125, 23.25, 0],
                              [-1.78125, 22.3125, 0],
                            ],
                            [null],
                            [null, [0, 24.28125, 0], [0, 22.3125, 0]],
                            [
                              null,
                              [0, 23.34375, 0],
                              [-0.1875, 23.53125, 0],
                              [-0.375, 23.625, 0],
                              [-0.65625, 23.625, 0],
                              [-0.84375, 23.53125, 0],
                              [-1.03125, 23.34375, 0],
                              [-1.125, 23.0625, 0],
                              [-1.125, 22.875, 0],
                              [-1.03125, 22.59375, 0],
                              [-0.84375, 22.40625, 0],
                              [-0.65625, 22.3125, 0],
                              [-0.375, 22.3125, 0],
                              [-0.1875, 22.40625, 0],
                              [0, 22.59375, 0],
                            ],
                            [null],
                            [
                              null,
                              [0.65625, 23.0625, 0],
                              [1.78125, 23.0625, 0],
                              [1.78125, 23.25, 0],
                              [1.6875, 23.4375, 0],
                              [1.59375, 23.53125, 0],
                              [1.40625, 23.625, 0],
                              [1.125, 23.625, 0],
                              [0.9375, 23.53125, 0],
                              [0.75, 23.34375, 0],
                              [0.65625, 23.0625, 0],
                              [0.65625, 22.875, 0],
                              [0.75, 22.59375, 0],
                              [0.9375, 22.40625, 0],
                              [1.125, 22.3125, 0],
                              [1.40625, 22.3125, 0],
                              [1.59375, 22.40625, 0],
                              [1.78125, 22.59375, 0],
                            ],
                            [null],
                            [
                              null,
                              [3, 24.28125, 0],
                              [2.8125, 24.28125, 0],
                              [2.625, 24.1875, 0],
                              [2.53125, 23.90625, 0],
                              [2.53125, 22.3125, 0],
                            ],
                            [null, [2.25, 23.625, 0], [2.90625, 23.625, 0]],
                            [null],
                            [
                              null,
                              [3.46875, 24.28125, 0],
                              [3.5625, 24.1875, 0],
                              [3.65625, 24.28125, 0],
                              [3.5625, 24.375, 0],
                              [3.46875, 24.28125, 0],
                            ],
                            [null, [3.5625, 23.625, 0], [3.5625, 22.3125, 0]],
                            [null],
                            [null, [4.3125, 23.625, 0], [4.3125, 22.3125, 0]],
                            [
                              null,
                              [4.3125, 23.25, 0],
                              [4.59375, 23.53125, 0],
                              [4.78125, 23.625, 0],
                              [5.0625, 23.625, 0],
                              [5.25, 23.53125, 0],
                              [5.34375, 23.25, 0],
                              [5.34375, 22.3125, 0],
                            ],
                            [null],
                            [
                              null,
                              [6, 23.0625, 0],
                              [7.125, 23.0625, 0],
                              [7.125, 23.25, 0],
                              [7.03125, 23.4375, 0],
                              [6.9375, 23.53125, 0],
                              [6.75, 23.625, 0],
                              [6.46875, 23.625, 0],
                              [6.28125, 23.53125, 0],
                              [6.09375, 23.34375, 0],
                              [6, 23.0625, 0],
                              [6, 22.875, 0],
                              [6.09375, 22.59375, 0],
                              [6.28125, 22.40625, 0],
                              [6.46875, 22.3125, 0],
                              [6.75, 22.3125, 0],
                              [6.9375, 22.40625, 0],
                              [7.125, 22.59375, 0],
                            ],
                            [null],
                            [null, [8.8125, 24.28125, 0], [8.8125, 22.3125, 0]],
                            [
                              null,
                              [8.8125, 23.34375, 0],
                              [8.625, 23.53125, 0],
                              [8.4375, 23.625, 0],
                              [8.15625, 23.625, 0],
                              [7.96875, 23.53125, 0],
                              [7.78125, 23.34375, 0],
                              [7.6875, 23.0625, 0],
                              [7.6875, 22.875, 0],
                              [7.78125, 22.59375, 0],
                              [7.96875, 22.40625, 0],
                              [8.15625, 22.3125, 0],
                              [8.4375, 22.3125, 0],
                              [8.625, 22.40625, 0],
                              [8.8125, 22.59375, 0],
                            ],
                            [null],
                          ],
                        },
                      ],
                    },
                    {
                      type: 'sketch',
                      content: [
                        {
                          type: 'disjointAssembly',
                          content: [
                            {
                              type: 'disjointAssembly',
                              content: [
                                {
                                  type: 'paths',
                                  tags: ['color/red'],
                                  paths: [
                                    [
                                      [21.000000000000004, 21, 0],
                                      [
                                        -21.000000000000004, 21.000000000000004,
                                        0,
                                      ],
                                      [
                                        -21.000000000000007,
                                        -20.999999999999996, 0,
                                      ],
                                      [21, -21.000000000000007, 0],
                                    ],
                                  ],
                                },
                              ],
                            },
                            {
                              type: 'paths',
                              tags: ['color/red'],
                              paths: [
                                [
                                  null,
                                  [-19.781250000000004, 24.28125, 0],
                                  [-20.718750000000004, 22.96875, 0],
                                  [-19.312500000000004, 22.96875, 0],
                                ],
                                [
                                  null,
                                  [-19.781250000000004, 24.28125, 0],
                                  [-19.781250000000004, 22.3125, 0],
                                ],
                                [null],
                                [
                                  null,
                                  [-18.750000000000004, 23.8125, 0],
                                  [-18.750000000000004, 23.90625, 0],
                                  [-18.656250000000004, 24.09375, 0],
                                  [-18.562500000000004, 24.1875, 0],
                                  [-18.375000000000004, 24.28125, 0],
                                  [-18.000000000000004, 24.28125, 0],
                                  [-17.812500000000004, 24.1875, 0],
                                  [-17.718750000000004, 24.09375, 0],
                                  [-17.625000000000004, 23.90625, 0],
                                  [-17.625000000000004, 23.71875, 0],
                                  [-17.718750000000004, 23.53125, 0],
                                  [-17.906250000000004, 23.25, 0],
                                  [-18.843750000000004, 22.3125, 0],
                                  [-17.531250000000004, 22.3125, 0],
                                ],
                                [null],
                                [null],
                                [
                                  null,
                                  [-15.468750000000004, 23.625, 0],
                                  [-14.437500000000004, 22.3125, 0],
                                ],
                                [
                                  null,
                                  [-14.437500000000004, 23.625, 0],
                                  [-15.468750000000004, 22.3125, 0],
                                ],
                                [null],
                                [null],
                                [
                                  null,
                                  [-11.437500000000002, 24.28125, 0],
                                  [-12.375000000000002, 22.96875, 0],
                                  [-10.968750000000002, 22.96875, 0],
                                ],
                                [
                                  null,
                                  [-11.437500000000002, 24.28125, 0],
                                  [-11.437500000000002, 22.3125, 0],
                                ],
                                [null],
                                [
                                  null,
                                  [-10.406250000000002, 23.8125, 0],
                                  [-10.406250000000002, 23.90625, 0],
                                  [-10.312500000000002, 24.09375, 0],
                                  [-10.218750000000002, 24.1875, 0],
                                  [-10.031250000000002, 24.28125, 0],
                                  [-9.656250000000002, 24.28125, 0],
                                  [-9.468750000000002, 24.1875, 0],
                                  [-9.375000000000002, 24.09375, 0],
                                  [-9.281250000000002, 23.90625, 0],
                                  [-9.281250000000002, 23.71875, 0],
                                  [-9.375000000000002, 23.53125, 0],
                                  [-9.562500000000002, 23.25, 0],
                                  [-10.500000000000002, 22.3125, 0],
                                  [-9.187500000000002, 22.3125, 0],
                                ],
                                [null],
                                [null],
                                [
                                  null,
                                  [-6.937500000000002, 23.625, 0],
                                  [-7.031250000000002, 23.53125, 0],
                                  [-6.937500000000002, 23.4375, 0],
                                  [-6.843750000000002, 23.53125, 0],
                                  [-6.937500000000002, 23.625, 0],
                                ],
                                [
                                  null,
                                  [-6.937500000000002, 22.5, 0],
                                  [-7.031250000000002, 22.40625, 0],
                                  [-6.937500000000002, 22.3125, 0],
                                  [-6.843750000000002, 22.40625, 0],
                                  [-6.937500000000002, 22.5, 0],
                                ],
                                [null],
                                [null],
                                [
                                  null,
                                  [-4.59375, 23.625, 0],
                                  [-4.59375, 22.6875, 0],
                                  [-4.5, 22.40625, 0],
                                  [-4.3125, 22.3125, 0],
                                  [-4.03125, 22.3125, 0],
                                  [-3.84375, 22.40625, 0],
                                  [-3.5625, 22.6875, 0],
                                ],
                                [
                                  null,
                                  [-3.5625, 23.625, 0],
                                  [-3.5625, 22.3125, 0],
                                ],
                                [null],
                                [
                                  null,
                                  [-2.8125, 23.625, 0],
                                  [-2.8125, 22.3125, 0],
                                ],
                                [
                                  null,
                                  [-2.8125, 23.25, 0],
                                  [-2.53125, 23.53125, 0],
                                  [-2.34375, 23.625, 0],
                                  [-2.0625, 23.625, 0],
                                  [-1.875, 23.53125, 0],
                                  [-1.78125, 23.25, 0],
                                  [-1.78125, 22.3125, 0],
                                ],
                                [null],
                                [null, [0, 24.28125, 0], [0, 22.3125, 0]],
                                [
                                  null,
                                  [0, 23.34375, 0],
                                  [-0.1875, 23.53125, 0],
                                  [-0.375, 23.625, 0],
                                  [-0.65625, 23.625, 0],
                                  [-0.84375, 23.53125, 0],
                                  [-1.03125, 23.34375, 0],
                                  [-1.125, 23.0625, 0],
                                  [-1.125, 22.875, 0],
                                  [-1.03125, 22.59375, 0],
                                  [-0.84375, 22.40625, 0],
                                  [-0.65625, 22.3125, 0],
                                  [-0.375, 22.3125, 0],
                                  [-0.1875, 22.40625, 0],
                                  [0, 22.59375, 0],
                                ],
                                [null],
                                [
                                  null,
                                  [0.65625, 23.0625, 0],
                                  [1.78125, 23.0625, 0],
                                  [1.78125, 23.25, 0],
                                  [1.6875, 23.4375, 0],
                                  [1.59375, 23.53125, 0],
                                  [1.40625, 23.625, 0],
                                  [1.125, 23.625, 0],
                                  [0.9375, 23.53125, 0],
                                  [0.75, 23.34375, 0],
                                  [0.65625, 23.0625, 0],
                                  [0.65625, 22.875, 0],
                                  [0.75, 22.59375, 0],
                                  [0.9375, 22.40625, 0],
                                  [1.125, 22.3125, 0],
                                  [1.40625, 22.3125, 0],
                                  [1.59375, 22.40625, 0],
                                  [1.78125, 22.59375, 0],
                                ],
                                [null],
                                [
                                  null,
                                  [3, 24.28125, 0],
                                  [2.8125, 24.28125, 0],
                                  [2.625, 24.1875, 0],
                                  [2.53125, 23.90625, 0],
                                  [2.53125, 22.3125, 0],
                                ],
                                [null, [2.25, 23.625, 0], [2.90625, 23.625, 0]],
                                [null],
                                [
                                  null,
                                  [3.46875, 24.28125, 0],
                                  [3.5625, 24.1875, 0],
                                  [3.65625, 24.28125, 0],
                                  [3.5625, 24.375, 0],
                                  [3.46875, 24.28125, 0],
                                ],
                                [
                                  null,
                                  [3.5625, 23.625, 0],
                                  [3.5625, 22.3125, 0],
                                ],
                                [null],
                                [
                                  null,
                                  [4.3125, 23.625, 0],
                                  [4.3125, 22.3125, 0],
                                ],
                                [
                                  null,
                                  [4.3125, 23.25, 0],
                                  [4.59375, 23.53125, 0],
                                  [4.78125, 23.625, 0],
                                  [5.0625, 23.625, 0],
                                  [5.25, 23.53125, 0],
                                  [5.34375, 23.25, 0],
                                  [5.34375, 22.3125, 0],
                                ],
                                [null],
                                [
                                  null,
                                  [6, 23.0625, 0],
                                  [7.125, 23.0625, 0],
                                  [7.125, 23.25, 0],
                                  [7.03125, 23.4375, 0],
                                  [6.9375, 23.53125, 0],
                                  [6.75, 23.625, 0],
                                  [6.46875, 23.625, 0],
                                  [6.28125, 23.53125, 0],
                                  [6.09375, 23.34375, 0],
                                  [6, 23.0625, 0],
                                  [6, 22.875, 0],
                                  [6.09375, 22.59375, 0],
                                  [6.28125, 22.40625, 0],
                                  [6.46875, 22.3125, 0],
                                  [6.75, 22.3125, 0],
                                  [6.9375, 22.40625, 0],
                                  [7.125, 22.59375, 0],
                                ],
                                [null],
                                [
                                  null,
                                  [8.8125, 24.28125, 0],
                                  [8.8125, 22.3125, 0],
                                ],
                                [
                                  null,
                                  [8.8125, 23.34375, 0],
                                  [8.625, 23.53125, 0],
                                  [8.4375, 23.625, 0],
                                  [8.15625, 23.625, 0],
                                  [7.96875, 23.53125, 0],
                                  [7.78125, 23.34375, 0],
                                  [7.6875, 23.0625, 0],
                                  [7.6875, 22.875, 0],
                                  [7.78125, 22.59375, 0],
                                  [7.96875, 22.40625, 0],
                                  [8.15625, 22.3125, 0],
                                  [8.4375, 22.3125, 0],
                                  [8.625, 22.40625, 0],
                                  [8.8125, 22.59375, 0],
                                ],
                                [null],
                              ],
                            },
                          ],
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          },
        },
      ];

      const run = async () => {
        const body = document.getElementsByTagName('body')[0];
        const notebookElement = await toDomElement(notebook);
        body.appendChild(notebookElement);
      };

      document.onreadystatechange = () => {
        if (document.readyState === 'complete') {
          run();
        }
      };
    </script>
  </body>
</html>
